### ISSUE TITLE
Hue ramp off not working

### ISSUE THREAD

**Issue by autoSteve** [autoSteve/acMqtt#8]
An issue has been introduced in HUE send receive that prevents a CBus ramp off working correctly.

**Comment by autoSteve** [comment]
Actually the HUE send receive script is quite broken for ramping. Works perfectly when no ramp is involved.

I see a re-write coming on...

A Philips Hue V2/V1 REST API hybrid is needed, because 'transitiontime' is seemingly not available yet in V2. But V2 is needed for the event stream. But the event stream is not 'snappy' with updates, so this makes timing ramps super difficult. This will require a re-think, likely requiring tossing messages to Hue and hoping to align status rather than waiting for verification that CBus and Hue are in agreement. A leap of faith? Probably got "**** up" written all over it... Maybe verify the 'final' level? (If that's possible.)

Beer. Sleep. Code. Repeat.

**Comment by autoSteve** [comment]
The Hue API is an interesting beast, whether V1 or V2.

I've got Cus ramps now instructing the bridge to use a 'transitiontime'. This is working perfectly so far, after much if/but/yeah/nah handling.

But, the thing with the API is that say I instruct a light to ramp from 127 to 0 then status updates periodically come in from the bridge as expected. Say 118, then 65, then 6, then zero. Cool. BUT if I then turn on the same light with the Hue app (which does not do transitiontime, probably for a good reason) then the light does come on, but at LEVEL 6, because that was the last status it knew about...

It is quite possibly one of the most half baked and frustrating APIs that I've worked with.

So in summary, if all commands come from CBus it all works as expected.

My next step is to attempt to code around this weirdness, probably with some kind of "last level" mechanism like I use for CBus<->HA. Essentially fire additional messages at the Hue should it get things "wrong".

I don't like complexity, but this script is turning into quite a complex beast.

Will commit an update with it as-is anyway.

**Comment by autoSteve** [comment]
Lastlevel is implemented, but it's not wirthout a quirk. I'm not sure if this is caused by the Philips eventstream implementation, or me, or comms issues.

Scenario:
1. Set a lamp to 50% in Hue app
2. Ramp it to off over four seconds in CBus
3. Attempt to turn the light on in Hue app. Occasionally it doesn't work as it should. For a while... and then it does. This happens  often but not always. (Just did five ramps off then turn on in a row and ... fine. The sixth took about 15-20 seconds for the final level to come in.)

What is happening in this corner case is that the API is refusing all commands to that light from the script until an eventstream message _finally_ comes in with the ultimate brightness level after completion of the 'transitiontime' (ramp) which will be 0.39%. Normally that final level comes in swiftly, but for some unknown reason at other times it does not. The impact is that the light does not return to its 'last level' as attempted to be set by the script because the API call just does nothing. No error. Nothing. And worse, the 'last level' then often gets set to 0.39%.Ugh.

This does not really impact me, as my primary use case for a ramp off is an "All off" button at bed time...

It may be that this is a Zigbee communication issue. The lamp is on a table in front of me, and it, the laptop and I are quite far away from the bridge, outside the house in a den. I intentionally develop this stuff at marginal communication range in the man cave, so it may be that the lamp I'm testing with is just retrying to get its status back to the bridge. This might also explain it possibly ignoring other commands, but an error message from the API instead of just silent nothing would be nice if that is the case.

When it's daylight again after a nap I'll place the lamp back beside the bed and re-test. Hopefully it's just dodgy comms.

Altering the CBus level during the ramp is also untested, as is probably more cases. But it works for me for now so it's a commit.

**Comment by autoSteve** [comment]
The quirk appears to be gone with better communication range. So closing as fixed.
